packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/VersionTracker.kt

```diff
@@ -32,14 +34,33 @@ import kotlinx.atomicfu.atomic
 internal class VersionTracker(private val owner: BaseRealmImpl, private val log: ContextLogger) {
     // Set of currently open realms. Storing the native pointer explicitly to enable us to close
     // the realm when the RealmReference is no longer referenced anymore.
-    private val intermediateReferences: AtomicRef<Set<Pair<RealmPointer, WeakReference<RealmReference>>>> = atomic(mutableSetOf())
+    private val intermediateReferences: AtomicRef<MutableSet<IntermediateReference>> =
```

Comment:
I actually think that this is from the old memory model, so don't think we need to "atominice" this anymore. And as long as you ensure synchronization elsewhere you could just operate on the mutable set. At least I think, so probably worth an experiment. If so, then `trackReference` could just add the single reference and  close expired could be trimmed to something like
```
intermediateReferences.removeAll { (pointer, ref) ->
    val reference = ref.get()
    return if (reference == null) {
        log.trace("$owner CLOSE-FREED ${RealmInterop.realm_get_version_id(pointer)}")
        RealmInterop.realm_close(pointer)
        true
    } else false
}
```

Reply:
I experimented removing the atomicity, but then it ran into race conditions. But that might be because I was initiating a new set. I will give it another try.

Reply:
Make it a regular set.

Reply:
??

Reply:
It is a note to remember to use a read-only set.